### PR TITLE
Handle reserved Kotlin keywords

### PR DESCRIPTION
**Pull request by asashour**
This is the second trial of https://github.com/flutter/flutter/pull/84973 which was reverted in #86512

The reason was `Template` signature has changed since the first PR was tested, and the time of merge.

The last commit shows the difference.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/src/template.dart [hunk 1]
```diff
@@ -322,6 +322,18 @@ class Template {
 
       if (sourceFile.path.endsWith(templateExtension)) {
         final String templateContents = sourceFile.readAsStringSync();
+        final String? androidIdentifier = context['androidIdentifier'] as String?;
+        if (finalDestinationFile.path.endsWith('.kt') && androidIdentifier != null) {
+          final List<String> segments = androidIdentifier
+              .split('.')
+              .toList();
```
Comment (shreyaskapale) [inline packages/flutter_tools/lib/src/template.dart:329]: I don't think toList() is required here. split() function returns a list.

Comment (asashour) [reply packages/flutter_tools/lib/src/template.dart:329]: Sure, obviously a newbie dart developer :)

File: packages/flutter_tools/lib/src/template.dart [hunk 2]
```diff
@@ -322,6 +322,16 @@ class Template {
 
       if (sourceFile.path.endsWith(templateExtension)) {
         final String templateContents = sourceFile.readAsStringSync();
+        final String? androidIdentifier = context['androidIdentifier'] as String?;
+        if (finalDestinationFile.path.endsWith('.kt') && androidIdentifier != null) {
+          final List<String> segments = androidIdentifier.split('.');
+          final List<String> reserved = <String>['when', 'in'];
+          final List<String> correctedSegments = segments.map(
+              (String segment) => reserved.contains(segment) ? '`$segment`' : segment
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/template.dart:330]: I think this is the wrong spot to perform the check. I would move this to create_base where the identifier is parsed.

For the kotlin reserved words, this should be in a top level constant with a link to documentation so we can keep it up to date.

Comment (asashour) [reply packages/flutter_tools/lib/src/template.dart:330]: The `androidIdentifier` is used for multiple purposes, e.g. creating the folder, in `android\app\src\main\kotlin` and for the package name inside the `MainActivity.kt`.

Putting the logic globally (inside the `create_base`), would make a folder named with \`in\` which is not what is needed AFAIHK.

The reserved keywords were put in a static top-level field.

File: packages/flutter_tools/lib/src/template.dart (hunk 1 above)
Comment (shreyaskapale) [reply packages/flutter_tools/lib/src/template.dart]: Haha, no worries happens sometimes.

**Review by jonahwilliams — COMMENTED** [review]
Oh I see, you're escaping them.

In that case I would prefer you pull that logic out into a dedicate helper method

File: packages/flutter_tools/lib/src/template.dart [hunk 3]
```diff
@@ -139,6 +139,8 @@ class Template {
   static const String copyTemplateExtension = '.copy.tmpl';
   static const String imageTemplateExtension = '.img.tmpl';
   static const String testTemplateExtension = '.test.tmpl';
+  // https://kotlinlang.org/docs/keyword-reference.html
+  static const List<String> reservedKotlinKeywords = <String>['when', 'in'];
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/template.dart:143]: make this a top level const `kReservedKotlinKeywords` and document what it is used for too

Comment (asashour) [reply packages/flutter_tools/lib/src/template.dart]: Done, thanks.

File: packages/flutter_tools/test/general.shard/android/template_test.dart [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import 'package:flutter_tools/src/template.dart';
+
+import '../../src/common.dart';
+
+void main() {
+
+  testWithoutContext('kotlin reserved keywords', () {
+    final FileSystem fileSystem = MemoryFileSystem.test();
+    final BufferLogger logger = BufferLogger.test();
+    final Directory rootDir = fileSystem.systemTempDirectory.createTempSync('flutter_template_test.');
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/android/template_test.dart:18]: you don't need to create a temp dir if you're using a memory file system, you just use currentDirectory

**Review by jonahwilliams — APPROVED** [review]
LGTM with nit

**Pull request merged**